### PR TITLE
Add referee display layer

### LINKED ISSUE #3289: Visualize ball placement point in Thunderscope field widget
### Description of the task

The ball placement point should be visualized in the Thunderscope field widget. Implementation in #3282 

The implementation work is mostly done, so this issue is mainly for porting over these changes to our master branch + cleaning up the code. Read through the linked PR and try running the code on your machine.

### Acceptance criteria

- [ ] Add a "Referee" layer to the field widget and visualize the ball placement point in the layer 
- [ ] Visualization should include a circle centered at the ball placement point with a radius of 0.15 m representing the allowed tolerance for ball placement
- [ ] Circle should be shaded green/red depending on whether the ball is in the placement zone or not
- [ ] Test by sending Game Controller commands

### LINKED ISSUE #3333: Overlay the current game state and game controller command on top of the field widget
### Description of the task

We can currently see the GameState and Command from the gamecontroller on the "RefereeInfoWidget" in Thunderscope. It would be nice to see both the state and command over the field widget.

Possible implementation:
1. Add a new gl layer here to display it on top of the field (https://github.com/UBC-Thunderbots/Software/tree/master/src/software/thunderscope/gl/layers)
2. Register a `ThreadSafeBuffer` for `Referee` messages and parse them (see [`refereeinfo_widget.py`](https://github.com/UBC-Thunderbots/Software/blob/master/src/software/thunderscope/play/refereeinfo_widget.py) to see how this is done)
3. Display the [`GameState.Type`](https://github.com/UBC-Thunderbots/Software/blob/bc475ca6932cd3335ec82477d17f10b8eaa710d9/src/proto/ssl_gc_state.proto#L54) and [`Command.Type`](https://github.com/UBC-Thunderbots/Software/blob/bc475ca6932cd3335ec82477d17f10b8eaa710d9/src/proto/ssl_gc_state.proto#L33) on the field widget (as strings)

### Acceptance criteria

- [ ] Run Thunderscope and view gamecontroller game state and command on field widget

### PR DESCRIPTION
**Pull request by Lmh-java**
### Description
Add a referee info layer to display game state and command.

Testing steps:
1. turn on "Referee Info" option in the "Layers" dropdown
2. Expect to see info on the top right corner

### Testing Done
- Thunderscope runs
- CI

### Resolved Issues
resolves #3333 
resolves #3289

### Demo
<details>
    <summary>Dropdown menu to enable layer</summary>
         image
</details>

<details>
    <summary>Game state and command display</summary>
         image
</details>

<details>
    <summary>Ball placement visual (ball is not placed correctly)</summary>
         image
</details>

<details>
    <summary>Ball placement visual (ball is placed correctly)</summary>
        image
</details>

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue

### CODE REVIEW

File: src/software/thunderscope/gl/graphics/gl_label.py [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    def __init__(
+        self,
+        parent_item: Optional[GLGraphicsItem] = None,
+        font: QFont = QFont("Roboto", 8),
+        text_color: QColor = Colors.PRIMARY_TEXT_COLOR,
+        offset: tuple[int, int] = (0, 0),
+        text: str = "",
+    ) -> None:
+        """Initialize the GLGradientLegend
```
Comment (williamckha) [inline src/software/thunderscope/gl/graphics/gl_label.py]:
```suggestion
        """Initialize the GLLabel
```

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        self.gamestate_type_text: Optional[GLLabel] = None
+        self.command_type_text: Optional[GLLabel] = None
+
+    def refresh_graphics(self) -> None:
+        """Update displays in the layer"""
+        referee_proto = self.referee_vis_buffer.get(block=False, return_cached=False)
+        if not referee_proto:
+            return
+
+        referee_msg_dict = MessageToDict(referee_proto)
```
Comment (williamckha) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py:127]: You actually don't need `MessageToDict` (I'm not sure why we use it in some of our code). You can just access the fields directly e.g. `referee_proto.stage`

**Review by williamckha — COMMENTED** [review]
🔥 

Would you also be able to incorporate #3289 into this PR? There's already another PR that never got merged which implements it.

**Comment by Lmh-java** [comment]
> 🔥
> 
> Would you also be able to incorporate #3289 into this PR? There's already another PR that never got merged which implements it.

Yes, sure. 👍

**Comment by Lmh-java** [comment]
Just a little bit clarification here before I get started, is ball placement visulization supposed to be in the same layer with referee info?

In #3289, that's what being said, but in the associated PR, it seems to be added to a different layer.

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 2 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py:127]: After testing, I think in this case I might want to keep `MessageToDict`, since directly acessing `referee_proto.stage` only gives me the index number (instead of the text)

**Comment by williamckha** [comment]
> Just a little bit clarification here before I get started, is ball placement visulization supposed to be in the same layer with referee info?
> 
> In #3289, that's what being said, but in the associated PR, it seems to be added to a different layer.

I would put the ball placement visualization in the Referee Info layer you created

File: src/software/thunderscope/gl/graphics/gl_label.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+        text: str = "",
+    ) -> None:
+        """Initialize the GLLabel
+
+        :param parent_item: The parent item of the graphic
+        :param font: The font using to render the text
+        :param text_color: The color for rendering the text.
+        :param offset: The offset (x, y) from the viewport left and top edge
+                       to use when positioning the label.
+                       If x is negative then the x offset is |x| pixels from
```
Comment (sauravbanna) [inline src/software/thunderscope/gl/graphics/gl_label.py:29]: nit: is this necessary? feels like its a bit clearer to just require the offsets to be positive numbers always. from the perspective of someone using this method in the future, the sign flipping feels a bit roundabout

Comment (sauravbanna) [reply src/software/thunderscope/gl/graphics/gl_label.py:29]: not a huge deal tho

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+    # text color for count down
+    COUNT_DOWN_TEXT_COLOR = Colors.RED
+
+    @staticmethod
+    def is_point_in_circle(
+        point: tuple[float | int, float | int],
+        center: tuple[float | int, float | int],
+        radius: float | int,
+    ) -> bool:
+        """Returns true if the point is in the circle.
```
Comment (sauravbanna) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: nit: wouldn't it be better for this method to take in 2 points instead? it looks like where you're using it, you're manually converting Point objects to this tuple. Most places in our code deal with Points instead of tuples so would be useful for that as well. Also would make the method signature a bit cleaner lol

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+                color=self.COUNT_DOWN_TEXT_COLOR,
+            )
+
+        if not ball_placement_vis_proto and not self.ball_placement_point_hidden:
+            self.placement_tolerance_graphic.hide()
+            self.placement_target_graphic.hide()
+            self.robot_avoid_circle_graphic.hide()
+            self.ball_placement_countdown_graphic.hide()
+
+            self.ball_placement_point_hidden = True
```
Comment (sauravbanna) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: nit: better name for this would `self.ball_placement_in_progress`, since that what it really means

i know this was my code lol but it was during robocup so not the cleanest

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+            return
+
+        if ball_placement_vis_proto:
+            # move the ball placement graphics to the new point
+            new_placement_point = ball_placement_vis_proto.ball_placement_point
+
+            # update the color of the target circle according to the position of the ball.
+            # if the ball lies inside the tolerance circle, the circle will be green, otherwise red.
+            ball_state = self.cached_world.ball.current_state
+            if not self.ball_placement_point_hidden:
```
Comment (sauravbanna) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: nit: would be better to pull out each branch of this if statement into 2 helper methods

so lines 131-146 in one, and lines 148-186 in one

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+                        0,
+                    ],
+                )
+                self.ball_placement_countdown_graphic.show()
+
+                self.placement_start_time = time.time()
+                self.ball_placement_point_hidden = False
+
+            # shrinking or expanding placement target graphic
+            if self.shrink_target:
```
Comment (sauravbanna) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: nit: you can move lines 188-207 into the if statement branch starting at line 131

basically if there isn't a ball placement in progress, you initialize everything first. if there is one in progress, you do all the updating logic

Comment (sauravbanna) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: and also may be cleaner to pull out this chunk into a helper too with a descriptive name

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+
+        if not self.ball_placement_countdown_graphic:
+            self.ball_placement_countdown_graphic = GLTextItem(
+                parentItem=self,
+                font=QtGui.QFont("Roboto", 7, weight=700),
+                color=self.COUNT_DOWN_TEXT_COLOR,
+            )
+
+        if not ball_placement_vis_proto and not self.ball_placement_point_hidden:
+            self.placement_tolerance_graphic.hide()
```
Comment (sauravbanna) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: nit: same with this chunk, can be a helper

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 9]
```diff
@@ -0,0 +1,249 @@
...
+        referee_proto = self.referee_vis_buffer.get(block=False, return_cached=False)
+        if not referee_proto:
+            return
+
+        referee_msg_dict = MessageToDict(referee_proto)
+        if not referee_msg_dict:
+            return
+        self.cached_referee_info = referee_msg_dict
+
+        if not self.gamestate_type_text:
```
Comment (sauravbanna) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py:132]: nit: these 2 if statements are pretty identical, could probably pull them out to a new method

**Review by sauravbanna — DISMISSED** [review]
looks great! really good job on incorporating (+ improving!) my ball placement impl, just some minor nits

File: src/software/thunderscope/gl/graphics/gl_label.py (hunk 3 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/graphics/gl_label.py:29]: Lol I followed the design of other components on the same level for consistency. But sure, I will clean this up to make it more understandable.

Comment (Lmh-java) [reply src/software/thunderscope/gl/graphics/gl_label.py:29]: After some testing here, I found out that the easier way to place the label on the top right corner is just to keep this wired sign flipping, because it takes some code to fetch the width of the viewport before passing into this function. Instead, we can get viewport by whatever calls this function, so we probably need to use a negative sign to denote starting from the right side.

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 9 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py:132]: These two if statements are writting to class fields, I can parametrize relevant parameters, but I don't know how to parametrize class level field so everytime we call this method we write to a different file. It seems taking a lot of code...

**Comment by Andrewyx** [comment]
Can you verify that this works with the recent preliminary arm PR #3316 (ie. pull upstream)? You may need to run `./setup_software,sh` again from the env setup folder.

**Comment by Lmh-java** [comment]
> Can you verify that this works with the recent preliminary arm PR #3316 (ie. pull upstream)? You may need to run `./setup_software,sh` again from the env setup folder.

@Andrewyx Reinstalled env and rebased, it worked on my side.

File: src/shared/constants.h [hunk 10]
```diff
@@ -79,6 +79,14 @@ static const double BALL_MAX_RADIUS_METERS = 0.0215;
 // cover more than 20% of the ball
 static const double MAX_FRACTION_OF_BALL_COVERED_BY_ROBOT = 0.2;
 
+// The radius of a circle region where ball placement is acceptable (in meters).
+static const double BALL_PLACEMENT_TOLERANCE_RADIUS_METERS = 0.15;
+// The radius of the outer region where robots are not allowed to be during ball
+// placement (in meters)
+static const double BALL_PLACEMENT_ROBOT_AVOID_RADIUS_METERS = 0.5;
+// The time limit for ball placement in seconds
+static const int BALL_PLACEMENT_TIME_LIMIT_S = 30;
```
Comment (Andrewyx) [inline src/shared/constants.h]: Interestingly enough, it seems that our `constants.h` file is wildly inconsistent with its use of `static const` vs `constexpr`. From my knowledge, there is no real performance difference here aside from `constexpr` potentially preventing others in the future from assigning a non-compile time constant value to these variables. 

TLDR: ask the others about as I am not sure why this is the case.

File: src/software/ai/hl/stp/play/ball_placement/ball_placement_play.cpp [hunk 11]
```diff
@@ -17,7 +18,18 @@ void BallPlacementPlay::getNextTactics(TacticCoroutine::push_type &yield,
 
 void BallPlacementPlay::updateTactics(const PlayUpdate &play_update)
 {
-    fsm.process_event(BallPlacementPlayFSM::Update(control_params, play_update));
+    auto event = BallPlacementPlayFSM::Update(control_params, play_update);
+    fsm.process_event(event);
+
+    auto placement_point = event.common.world_ptr->gameState().getBallPlacementPoint();
```
Comment (Andrewyx) [inline src/software/ai/hl/stp/play/ball_placement/ball_placement_play.cpp]: nit: might be better to forgo auto and use `std::optional<Point>` here since the point can be `null_opt` type which explains the necessity of the null check below.

**Comment by Andrewyx** [comment]
LGTM great work! Just some small nits.

File: src/shared/constants.h (hunk 10 above)
Comment (Lmh-java) [reply src/shared/constants.h]: As the info I have learned just then, `constexpr` might be used in this case, but I will confirm with others.

`static const` seems to be used more in class-level constants. If we use `constexpr`, it allows complie-time optimization.

If confirmed and refactor needed, do you mind if I do this refactor in a separate ticket, so we don't mess up other features.

Comment (Andrewyx) [reply src/shared/constants.h]: Sure, sounds good.

IMO `constexpr` and `static const` in this case should not technically have a difference in optimization in this case as most of the `static const` values we have can be determined at compile-time which results in the optimization process being carried through anyways (to my understanding). This would mean the main advantage of switching would purely be to deter others in the future from using non-compile time derivable values (as `constexpr` would throw an error). You may want to refer to the link below for a clearer explanation:

https://stackoverflow.com/questions/41125651/constexpr-vs-static-const-which-one-to-prefer

Comment (williamckha) [reply src/shared/constants.h]: Yes, we could open a new issue to make all these constants `constexpr`. The main benefit would be so that we can use these constants in other `constexpr` variables and use them in template args, array declarations, etc.

File: src/software/ai/hl/stp/play/ball_placement/ball_placement_play.cpp (hunk 11 above)
Comment (Lmh-java) [reply src/software/ai/hl/stp/play/ball_placement/ball_placement_play.cpp]: Done

File: src/shared/constants.h (hunk 10 above)
Comment (Lmh-java) [reply src/shared/constants.h]: Done, fixed all 3 constants included in this PR with `constexpr`. Will send another PR on refactoring others after this is merged.

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 12]
```diff
@@ -0,0 +1,269 @@
...
+    BALL_PLACEMENT_TOLERANCE_VISUALIZATION_COLOR_ON = Colors.GREEN
+    # outline color of the tolerance region for ball placement (when the ball is outside)
+    BALL_PLACEMENT_TOLERANCE_VISUALIZATION_COLOR_OFF = Colors.RED
+    # outline color of the target mark
+    BALL_PLACEMENT_TARGET_VISUALIZATION_COLOR = Colors.FIELD_LINE_COLOR
+    # text color for count down
+    COUNT_DOWN_TEXT_COLOR = Colors.RED
+
+    @staticmethod
+    def is_point_in_circle(
```
Comment (itsarune) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: this helper isn't actually necessary, we've binded this function in python from C++

[C++ impl](https://github.com/UBC-Thunderbots/Software/blob/10bded3efa3780478ae10a76840a5edf17abb508/src/software/geom/algorithms/contains.h#L19)
[Python bindings](https://github.com/UBC-Thunderbots/Software/blob/10bded3efa3780478ae10a76840a5edf17abb508/src/software/python_bindings.cpp#L313)

You can call this via `tbots_cpp.contains(circle, point)`

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 9 above)
Comment (itsarune) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py:132]: you can probably parametrize a class-level member with a `lambda` but that might be overkill since there's only 2

**Review by itsarune — COMMENTED** [review]
nice work, just one small thing I noticed

File: src/software/thunderscope/widget_setup_functions.py [hunk 13]
```diff
@@ -141,6 +145,7 @@ def setup_gl_widget(
     gl_widget.add_layer(validation_layer)
     gl_widget.add_layer(trail_layer, False)
     gl_widget.add_layer(debug_shapes_layer, True)
+    gl_widget.add_layer(referee_layer, False)
```
Comment (itsarune) [inline src/software/thunderscope/widget_setup_functions.py]: I think this layer should be on by default

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 14]
```diff
@@ -0,0 +1,269 @@
...
+            self.placement_target_graphic.set_radius(
+                self.placement_target_graphic.radius + 0.01
+            )
+            self.shrink_target = (
+                self.placement_target_graphic.radius
+                >= BALL_PLACEMENT_TOLERANCE_RADIUS_METERS
+            )
+
+        # update the count-down graphics
+        time_left = max(
```
Comment (itsarune) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: this should be guarded with `if self.cached_referee_info`

Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: Done

File: src/software/thunderscope/widget_setup_functions.py (hunk 13 above)
Comment (Lmh-java) [reply src/software/thunderscope/widget_setup_functions.py]: Done

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 12 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: Done

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+                + BALL_PLACEMENT_ROBOT_AVOID_RADIUS_METERS / 2,
+                self.ball_placement_point.y_meters
+                + BALL_PLACEMENT_ROBOT_AVOID_RADIUS_METERS
+                + 0.1,
+                0,
+            ],
+        )
+        self.ball_placement_countdown_graphic.show()
+
+        self.placement_start_time = time.time()
```
Comment (itsarune) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: AFAICT: we don't seem to use `placement_start_time` anywhere

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 16]
```diff
@@ -0,0 +1,250 @@
...
+    def __update_ball_placement(self) -> None:
+        """Update ball placement visuals"""
+        ball_placement_vis_proto = self.ball_placement_vis_buffer.get(
+            block=False, return_cached=False
+        )
+
+        if not self.ball_placement_countdown_graphic:
+            self.ball_placement_countdown_graphic = GLTextItem(
+                parentItem=self,
+                font=QtGui.QFont("Roboto", 7, weight=700),
```
Comment (itsarune) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: Nit: It seems like we've standardized "Roboto" as our font of choice. Can you abstract it into a constant in `constants.py` and replace all uses?

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 15 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: Good call! Removed

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 16 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: Done

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 17]
```diff
@@ -0,0 +1,250 @@
...
+        if ball_placement_vis_proto:
+            new_placement_point = ball_placement_vis_proto.ball_placement_point
+            if not self.ball_placement_in_progress:
+                # initialize the visuals
+                self.ball_placement_tolerance_circle = tbots_cpp.Circle(
+                    tbots_cpp.createPoint(new_placement_point),
+                    BALL_PLACEMENT_TOLERANCE_RADIUS_METERS,
+                )
+                self.__display_ball_placement_visuals(new_placement_point)
+                self.ball_placement_in_progress = True
```
Comment (itsarune) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: why did you decide that `__display_ball_placement_visuals` would set `self.ball_placement_in_progress` to `False`? and then revert it here?

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py [hunk 18]
```diff
@@ -0,0 +1,250 @@
...
+        self.placement_target_graphic.show()
+
+        self.robot_avoid_circle_graphic.set_position(
+            self.ball_placement_point.x_meters,
+            self.ball_placement_point.y_meters,
+        )
+        self.robot_avoid_circle_graphic.show()
+
+        self.ball_placement_countdown_graphic.setData(
+            text=f"{BALL_PLACEMENT_TIME_LIMIT_S}s",
```
Comment (itsarune) [inline src/software/thunderscope/gl/layers/gl_referee_info_layer.py:230]: seems like `BALL_PLACEMENT_TIME_LIMIT_S` is undefined

**Review by itsarune — COMMENTED** [review]
looks good, 2 more minor things

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 18 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py:230]: I think it is defined in `src/shared/constants.h` of my PR

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 17 above)
Comment (Lmh-java) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py]: Good catch! Now, directly set the flag

File: src/software/thunderscope/gl/layers/gl_referee_info_layer.py (hunk 18 above)
Comment (itsarune) [reply src/software/thunderscope/gl/layers/gl_referee_info_layer.py:230]: Oh yes you're right!

**Review by itsarune — DISMISSED** [review]
LGTM!

**Review by williamckha — DISMISSED** [review]
LGTM, pending merge conflicts

**Comment by Lmh-java** [comment]
@itsarune @williamckha Merged with master, manual tested and everything LGTM. Please review

**Review by williamckha — APPROVED** [review]
sorry this has been in review for so long 😅

**Comment by Lmh-java** [comment]
Thanks William! 🍻

**Review by itsarune — APPROVED** [review]
full send :airplane: :airplane:
